lab4/src/main/java/agh/ics/oop/RectangularMap.java

```diff
@@ -0,0 +1,59 @@
+package agh.ics.oop;
+
+import java.util.HashMap;
+
+class RectangularMap implements IWorldMap {
+
+    // hashmap which keys are position (as vector) and values are Animal
+    private final HashMap<Vector2d, Animal> animalPositions;
+
+    private final Vector2d UPPER_RIGHT;
+    private final Vector2d LOWER_LEFT = new Vector2d(0, 0);
+    private final MapVisualizer visualizer;
+
+
+    public RectangularMap(int width, int height) {
+        UPPER_RIGHT = new Vector2d(width, height);
+        this.visualizer = new MapVisualizer(this);
+        this.animalPositions = new HashMap<>();
+    }
+
+
+    @Override
+    public boolean canMoveTo(Vector2d position) {
+        return !animalPositions.containsKey(position) && position.precedes(UPPER_RIGHT) && position.follows(LOWER_LEFT);
+    }
+
+    @Override
+    public boolean place(Animal animal) {
+        if(!animalPositions.containsKey(animal.getPosition())) {
+            animalPositions.put(animal.getPosition(), animal);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void changePosition(Vector2d oldPosition, Vector2d newPosition) {
```

Comment:
przy rozwiązaniu z mapą zamiast listy dobrze byłoby przenieść całą odpowiedzialność za pozycje w jedno miejsce - czyli pozbyć się atrybutu pozycji z klasy `Animal` (trzymanie tej samej informacji w dwóch miejscach zawsze kończy się źle). Ale to jeszcze będziemy kombinować na laborce z map ;)